File: server/Models/Patient.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using MongoDB.Bson.Serialization.Attributes;
using server;

namespace Server.Core.Models
{
    [BsonIgnoreExtraElementsAttribute]
    public class Patient : BasicPerson, ITeamEntity
    {
        public string Id { get; set; }
        public string TeamId { get; set; }
        public DateTime CreatedAt { get; set; }
        public DateTime UpdatedAt { get; set; }

        public string MedicalInsurance { get; set; }

        public BasicPerson AccountablePerson { get; set; } = new BasicPerson();
        public BasicPerson Mother { get; set; } = new BasicPerson();
        public BasicPerson Father { get; set; } = new BasicPerson();

        public List<SocialProfile> SocialProfiles { get; set; } = new List<SocialProfile>();
        public List<MedicalReceipt> MedicalReceipts { get; set; } = new List<MedicalReceipt>();
        public string MedicalObservations { get; set; }
        
        public AppointmentInfo AppointmentInfo { get; set; } = new AppointmentInfo();
        public string Status { get; set; }

        [DisplayFormat(DataFormatString = "yyyy-MM-ddThh:mm")]
        public DateTime LastVisit { get; set; }

        public string Gender { get; set; }
        public Address Address { get; set; } = new Address();
        public Address CommercialAddress { get; set; } = new Address();

        [DisplayFormat(DataFormatString = "yyyy-MM-dd")]
        public DateTime BirthDate { get; set; }

        public string MaritalStatus { get; set; }
        public string Spouse { get; set; }
        public string Notes { get; set; }

    }

    public class MedicalReceipt 
    {
        [DisplayFormat(DataFormatString = "yyyy-MM-ddThh:mm")]
        public DateTime CreatedAt { get; set; }

        public string Type { get; set; }
        public string Description { get; set; }
    }

    public class AppointmentInfo 
    {
        public List<DescriptionInfo> Motivations { get; set; } = new List<DescriptionInfo>();
        public List<DescriptionInfo> Indications { get; set; } = new List<DescriptionInfo>();
        public List<AppointmentVisit> Visits { get; set; } = new List<AppointmentVisit>();
    }

    public class AppointmentVisit 
    {
        [DisplayFormat(DataFormatString = "yyyy-MM-ddThh:mm")]
        public DateTime CreatedAt { get; set; }

        [DisplayFormat(DataFormatString = "yyyy-MM-ddThh:mm")]
        public DateTime VisitDate { get; set; }

        public string Status { get; set; }
        public string Description { get; set; }
    }

    public class BasicPerson 
    {
        public string Name { get; set; }
        public string TaxIdNumber { get; set; }
        public string RegistrationNumber { get; set; }
        public string Email { get; set; }        
        public string JobTitle { get; set; }
        public List<PhoneNumber> PhoneNumbers { get; set; } = new List<PhoneNumber>();

        [BsonIgnore]
        public string EmailHash => MD5Util.CreateMD5(Email);
    }

    public class Address 
    {
        public string Type { get; set; }
        public string PostalCode { get; set; }
        public string Street { get; set; }
        public string StreetComplement { get; set; }
        public string City { get; set; }
        public string State { get; set; }
        public string Country { get; set; }
        public string Neighbourhood { get; set; }
    }

    public class PhoneNumber 
    {
        public string Type { get; set; }
        public string Number { get; set; }
    }

    public class SocialProfile 
    {
        public string Type { get; set; }
        public string Description { get; set; }
    }

    public class DescriptionInfo 
    {
        public string Type { get; set; }
        public string Description { get; set; }
    }

    public class PatientListViewModel 
    {
        public string Id { get; set; }
        public string Name { get; set; }
        public string TaxIdNumber { get; set; }
        public string MedicalInsurance { get; set; }
        public DateTime LastVisit { get; set; }
        public string Email { get; set; }
        
        [BsonIgnore]
        public string EmailHash => MD5Util.CreateMD5(Email);
    }

    public class PatientImportViewModel
    {
        public string Name { get; set; }
        public string Email { get; set; }
        public string MedicalInsurance { get; set; }
        public string AccountablePerson { get; set; }
        public string BirthDate { get; set; }
        public string LastAppointment { get; set; }
        public string PhoneNumber { get; set; }
        public string Address { get; set; }

        public Patient ToPatient() {
            DateTime birthDate;
            DateTime lastAppointment;
            var birthDateOk = DateTime.TryParse(this.BirthDate, out birthDate);
            var lastAppointmentOk = DateTime.TryParse(this.LastAppointment, out lastAppointment);
            return new Patient {
                Name = this.Name,
                Email = this.Email,
                MedicalInsurance = this.MedicalInsurance,
                AccountablePerson = new BasicPerson { Name = this.AccountablePerson },
                BirthDate = birthDateOk ? birthDate : default(DateTime),
                LastVisit  = lastAppointmentOk ? lastAppointment : default(DateTime),
                PhoneNumbers = new List<PhoneNumber> { new PhoneNumber { Type = "Contato", Number = this.PhoneNumber } },
                Address = new Address { Street = this.Address }
            };
        }
    }
    
}

File: mednuvem-web/src/providers/patient.ts
import { CalendarEvent } from 'angular-calendar';
import { startOfDay, addHours } from 'date-fns';

export class Team {
  id: string;
  createdAt: Date;
  updatedAt: Date;
  userId: string;
  name: string;
  members: TeamMember[] = [];
  messages: TeamChatMessage[] = [];
}

export class TeamMember {
  userId: string;
  role: string;
  name: string;
  email: string;
  isOnline: boolean;
}

export class TeamChatMessage {
  sentAt: Date;
  userId: string;
  email: string;
  message: string;
  userName: string;
  userEmail: string;
}

export class BasicPerson {
  name: string;
  taxIdNumber: string;
  registrationNumber: string;
  email: string;
  emailHash: string;
  jobTitle: string;
  phoneNumbers: PhoneNumber[] = [];
}

export class Patient extends BasicPerson {
  id: string;
  userId: string;
  createdAt: Date;
  updatedAt: Date;
  birthDate: Date;
  gender: string;
  medicalInsurance: string;
  accountablePerson: BasicPerson = new BasicPerson();
  mother: BasicPerson = new BasicPerson();
  father: BasicPerson = new BasicPerson();
  socialProfiles: SocialProfile[] = [];
  medicalReceipts: MedicalReceipt[] = [];
  medicalObservations: string;

  status: string;
  lastVisit: Date;
  appointmentInfo: AppointmentInfo = new AppointmentInfo();

  address: Address = new Address();
  commercialAddress: Address = new Address();
  maritalStatus: string;
  spouse: string;
  notes: string;
}

export class Address {
	type: string;
	postalCode: string;
	street: string;
	streetComplement: string;
	city: string;
	state: string;
	country: string;
	neighbourhood: string;
}

export class PhoneNumber {
	type: string;
	number: string;
}

export class SocialProfile {
  type: string;
  description: string;
}

export class MedicalReceipt {
  type: string;
  description: string;
}

export class AppointmentInfo {
  motivations: DescriptionInfo[] = [];
  indications: DescriptionInfo[] = [];
  visits: AppointmentVisit[] = [];
}

export class DescriptionInfo {
  type: string;
  description: string;
}

export class AppointmentVisit {
  createdAt: Date;
  visitDate: Date;
  status: string;
  description: string;
}

export class PatientCalendarEvent {
  color: EventColor;
  actions?: EventAction[];
  allDay?: boolean;
  cssClass?: string;
  resizable?: { beforeStart?: boolean; afterEnd?: boolean; };
  draggable?: boolean;

  id?: string;
  start: string = startOfDay(new Date()).toISOString();
  end?: string = addHours(startOfDay(new Date()), 1).toISOString();
  title: string;
  type?: string = 'Consulta';
  description?: string;
  status?: string;
}

export interface EventColor {
  primary: string;
  secondary: string;
}
export interface EventAction {
  label: string;
  cssClass?: string;
  onClick({event}: {
    event: PatientCalendarEvent;
  }): any;
}
export interface CalendarEventTimesChangedEvent {
  event: PatientCalendarEvent;
  newStart: string;
  newEnd?: string;
}
